core/domain/exp_domain.py

```diff
@@ -474,6 +474,246 @@ def __init__(self, change_list):
+
+        Returns:
+            list(tuple(str, state_domain.State)). State details corresponding to
+            the given range of versions.
+        """
+        version_range = python_utils.RANGE(
+            max(self._version_start, version_start),
+            min(self._version_end, version_end))
+        return [self[v] for v in version_range]
+
+
+class ExplorationStateHistory(python_utils.OBJECT):
+    """Organizes the history of changes made to each state of an exploration."""
+
+    def __init__(self, exps, exp_version_diffs):
+        """Constructs the exploration's state history.
+
+        Args:
+            exps: list(Exploration). The exploration's representation, ordered
+                by version. Items should begin at version 1 and contain no gaps.
+            exp_version_diffs: list(ExplorationVersionsDiff). The specific
+                changes made to the exploration at each corresponding version.
+        """
+        if not exps or not exp_version_diffs:
+            raise ValueError('Inputs must be non-empty')
+
+        if len(exps) != len(exp_version_diffs):
+            raise ValueError('len of inputs do not match')
+
+        exp_versions = [exp.version for exp in exps]
+        expected_versions = list(python_utils.RANGE(1, len(exps) + 1))
+        if exp_versions != expected_versions:
+            raise ValueError(
+                'Explorations must be sorted, begin from version 1, and have '
+                'no gaps; but got invalid input versions == [%s]' % (
+                    ', '.join(python_utils.UNICODE(v) for v in exp_versions)))
+
+        self._state_spans_history = []
+        for exp, diff in python_utils.ZIP(exps, exp_version_diffs):
+            new_state_spans = dict()
+            for name, state in exp.states.items():
+                if exp.version == 1 or name in diff.added_state_names:
+                    new_state_spans[name] = (
+                        StateVersionSpan(exp.version, name, state))
+                elif name not in diff.deleted_state_names:
+                    prev_state_spans, prev_name = (
+                        self._state_spans_history[-1],
+                        diff.new_to_old_state_names.get(name, name))
+                    new_state_spans[name] = (
+                        prev_state_spans[prev_name].extend_or_split(
+                            exp.version, name, state, diff))
+            self._state_spans_history.append(new_state_spans)
+
+    def get_state_span(self, exp_version, state_name):
+        """Returns the span of versions where the given state existed.
+
+        Args:
+            exp_version: int. Exploration version to query.
+            state_name: str. The name of the state at the given version.
+
+        Returns:
+            StateVersionSpan. The span of versions for the given state including
+            the given version.
+        """
+        return self._state_spans_history[exp_version - 1][state_name]
```

Comment:
May need to ensure state_name exists in the given version (or return None? If so, please update docstring accordingly.)

Reply:
Done.